### PR TITLE
[stable/traefik] Allow skipping secret creation for dns providers

### PR DESCRIPTION
**Pull request by kir4h**
#### What this PR does / why we need it:

This PR allows avoiding creation of secrets for DNS providers when using DNS acme challenge for generating LetsEncrypt certificates.

The use case is being able to read an existing generated secret for those required environment variables without the need of exposing those secrets to the values.yaml file. (e.g: An external step, could have created those variables beforehand and we just need to consume them)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

**Comment by kir4h** [comment]
Putting on hold as I realized missing items:

- README.md should specify expected secret names
- Some logic when referencing the secret should be reviewed (if we skip secret creation we should allow empty values for the environment variables)

**Comment by stealthybox** [comment]
> README.md should specify expected secret names

The canonical interface for this is parameterize the Secret name via values, so that the user has an explicit config point for referencing and overriding behavior.
This is necessary for things like Secret rotation that's not managed directly by a single Helm release.

**Comment by stealthybox** [comment]
+1 for this feature.
This is how I wanted the Traefik chart to behave before I started using it.

There have been instances where I was unable to check values into source control for use with existing pipelines because the DNS Secrets are currently required to be passed through the release values.

**Comment by stealthybox** [comment]
/lgtm

I'm not in `traefik/OWNERS` though.

**Comment by kir4h** [comment]
@juliens is there anything else I would need to take care of? First PR on Helm so not so familiar with the procedures

**Comment by kir4h** [comment]
@dtomcej Will you be able to review this one? Not sure whom to assign it to (tried initially with jbdoumenjou but assignment failed so I falled back to you based on latest PR approved for this chart)

Thanks!

File: stable/traefik/templates/deployment.yaml
Comment (dtomcej) [inline stable/traefik/templates/deployment.yaml:97]: Is this `or` needed? If we are iterating through the loop, shouldn't this always be true? leading to multiple mounts with the same keyref?

**Review by dtomcej — CHANGES_REQUESTED** [review]

File: stable/traefik/templates/deployment.yaml
Comment (kir4h) [reply stable/traefik/templates/deployment.yaml:97]: I think the `or` is needed and multiple mounts with the same keyref should not happen. I'm not very confident with templating so let me explain the logic.

#### Previous scenario (`{{- if $v }}`) 
If `dnsProvider.name` is set:
* We retrieve a dictionary of environment variable names and their values for the specified dnsProvider
* We iterate over those environment variables. For each environment variable:
  * If the value is set we create an `env` item with the environment variable name as key. This environment variable is retrieved from the secret `{{ template "traefik.fullname" $ }}-dnsprovider-config`, under a key with the same name as the environment variable
  * If the value is not set we skip it

#### PR scenario  ( `{{- if or $v $.Values.acme.dnsProvider.existingSecretName }}`)
If `dnsProvider.name` is set:
* We retrieve a dictionary of environment variable names and their values for the specified dnsProvider
* We iterate over those environment variables. For each environment variable:
  * And `env` item is created if either
    * value is set for that environment variable (previous behaviour)
    * `existingSecretName` is specified (that would mean we want to use an existingSecretName as source for those environment values)
  * If an `env` item is created for the environment variable, its value will be:
    * The value contained in the corresponding `existingSecret` (if `existingSecret` is set)
    * The value specified in values file otherwise

There would be an error of course if an `existingSecret` is specified, and that `existingSecret` doesn't contain the required keys but I would say that is the intended behaviour as those variables seem to be mandatory for using the provider.

Comment (kir4h) [reply stable/traefik/templates/deployment.yaml:97]: @dtomcej would you require additional clarification or should I mark the conversation as resolved?

**Comment by kir4h** [comment]
@dtomcej anything else I can help with for the approval?

**Comment by kir4h** [comment]
sorry @dtomcej, had to rebase to sync with upstream changes so lost the `lgtm` label

**Comment by kir4h** [comment]
@dtomcej I would need the lgtm label to proceed. My only change since your review is bumping the chart version because of https://github.com/helm/charts/pull/12609

**Pull request merged**
